www/nginx/src/opnsense/mvc/app/library/OPNsense/Nginx/LogParserBase.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+/*
+
+    Copyright (C) 2018-2020 Fabian Franz
+    Copyright (C) 2020 Manuel Faux
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+
+    1. Redistributions of source code must retain the above copyright notice,
+       this list of conditions and the following disclaimer.
+
+    2. Redistributions in binary form must reproduce the above copyright
+       notice, this list of conditions and the following disclaimer in the
+       documentation and/or other materials provided with the distribution.
+
+    THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+    INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+    AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+
+namespace OPNsense\Nginx;
+
+abstract class LogParserBase
+{
+    protected $result;
+
+    function __construct($file_name, $page, $per_page, $query)
+    {
+        $this->file_name = $file_name;
+        $this->page = $page;
+        $this->per_page = $per_page;
+        $this->query = $query;
+        $this->page_count = 0;
+        $this->total_lines = 0;
+        $this->query_lines = 0;
+        $this->result = array();
+
+        if (substr($this->file_name, -2) == "gz") {
+            $this->parse_gzfile();
+        }
+        else {
+            $this->parse_file();
+        }
+    }
+
+    /**
+     * Reverse read logfile, count lines, save lines which match filter and
+     * count matching lines.
+     */
+    private function parse_file()
+    {
+        $handle = @fopen($this->file_name, 'r');
+        if ($handle) {
+            $cnt = 0;
+            $total = 0;
+            for ($line = '', $pos = -2; fseek($handle, $pos, SEEK_END) !== -1; $pos--) {
```

Comment:
The PHP implementation of `SEEK_END` seems to slow the file reading dramatically (approx. 600ms vs. 70ms when reading the complete file linear).
As one nginx log file typically is not huge in size, maybe reading the complete file to memory is fine as well.